### PR TITLE
ZD-850 extend sdk with media

### PR DESCRIPTION
**Pull request by spotaichuk-smartling**

### CODE REVIEW

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 1]
```diff
@@ -532,6 +534,27 @@ public Iterable<Article> getArticleFromSearch(String locale, String searchTerm,
         return complete(submit(req("GET", tmpl), handleList(Article.class, "results")));
     }
 
+    public MediaRecord getMediaByName(String name) {
```
Comment (dimitrystd) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: IMO we should keep SDK simple. Your method includes business logic that is specific to the current task implementation. If we will decide to change the flow then we will need to update the SDK.
This method should take all possible parameters (search term, sorting, and pagination) and return the list as is. The actual pagination loop and logic should be in the connector.

File: src/main/java/org/zendesk/client/v2/model/hc/ArticleAttachments.java [hunk 2]
```diff
@@ -74,7 +77,6 @@ public void setUrl(String url) {
         this.url = url;
     }
 
-    @JsonProperty("article_id")
```
Comment (dimitrystd) [inline src/main/java/org/zendesk/client/v2/model/hc/ArticleAttachments.java]: Why did you remove this annotation for random properties, but not for all? I don't suggest deleting it everywhere. We should probably return it to ensure that it is consistent with the rest of the content in the file.

**Review by dimitrystd — CHANGES_REQUESTED** [review]

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 3]
```diff
@@ -532,6 +534,27 @@ public Iterable<Article> getArticleFromSearch(String locale, String searchTerm,
...
+        if (!optionalMediaRecord.isPresent()) {
+            while (response.getMeta().hasMore() && !optionalMediaRecord.isPresent()) {
+                tmpl = tmpl("/guide/medias?filter[name]={name}&page[size]={pageSize}&page[after]={cursor}")
+                        .set("name", name)
+                        .set("pageSize", apiMaxPageSize)
+                        .set("cursor", response.getMeta().getAfterCursor());
+                response = complete(submit(req("GET", tmpl), handle(MediaResponse.class)));
+                optionalMediaRecord = response.getRecords().stream().filter(record -> record.getName().equals(name)).findFirst();
+            }
+        }
```
Comment (PavelLoparev) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: I agree that this needs to be implemented in a client/service class code OR introduce wrapper method like `getAllMedia` with this logic.

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 4]
```diff
@@ -554,6 +577,16 @@ public List<ArticleAttachments> getAttachmentsFromArticle(String locale, Long ar
                 handleArticleAttachmentsList("article_attachments")));
     }
 
+    public ArticleAttachments saveAttachment(Long articleId, boolean inline, String guideMediaId, String locale) {
```
Comment (PavelLoparev) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: It's more like `createAttachment`. Minor comment

**Review by dimitrystd — APPROVED** [review]

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 5]
```diff
@@ -532,6 +536,20 @@ public Iterable<Article> getArticleFromSearch(String locale, String searchTerm,
...
+        StringBuilder sb = new StringBuilder("/guide/medias");
+        List<String> statements = new ArrayList<>();
+
+        statements.add(filterMap.entrySet().stream().map(entry -> String.format("filter%s=%s", entry.getKey(), entry.getValue())).collect(Collectors.joining("&")));
+        Optional.ofNullable(pageAfter).map(it -> String.format("page[after]=%s", it)).ifPresent(statements::add);
+        Optional.ofNullable(sorting).map(it -> String.format("sort=%s", it.getSortOrder().equals(DESCENDING) ? "-".concat(it.getSortBy()) : it.getSortBy())).ifPresent(statements::add);
+        if (!statements.isEmpty()) {
+            sb.append("?").append(statements.stream().filter(a -> !a.isEmpty()).collect(Collectors.joining("&")));
+        }
+        TemplateUri tmpl = tmpl(sb.toString());
```
Comment (PavelLoparev) [inline src/main/java/org/zendesk/client/v2/Zendesk.java]: maybe let's use some lib for constructing query strings. I think you would eventually need to URL encode query string values, no? like space to `%20` or `+` etc. I can imagine file name with spaces and any other characters that need to be url encoded

Comment (dimitrystd) [reply src/main/java/org/zendesk/client/v2/Zendesk.java]: Very good point. Users definitely will include crazy characters in image filenames.

Comment (spotaichuk-smartling) [reply src/main/java/org/zendesk/client/v2/Zendesk.java]: It should be encoded in the same way as it was before. Here we just forward some params to zendesk utility class

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 6]
```diff
@@ -2476,6 +2479,18 @@ public void deleteTranslation(Long translationId) {
                 handleStatus()));
     }
 
+    public ArticleAttachments createArticleAttachment(Long articleId, boolean inline, String guideMediaId, String locale) {
+        Map<String, Object> request = new HashMap<>();
+        request.put("inline", inline);
+        request.put("guide_media_id", guideMediaId);
+        request.put("locale", locale);
+
+        TemplateUri tmpl = tmpl("/help_center/articles/{articleId}/attachments")
```
Comment (PavelLoparev) [inline src/main/java/org/zendesk/client/v2/Zendesk.java:2488]: Url for creating target article attachment is `https://{{subdomain}}.zendesk.com/api/v2/help_center/{{targetLocale}}/articles/{{articleId}}/attachments`. locale in URL but not in the form params.

File: src/main/java/org/zendesk/client/v2/Zendesk.java [hunk 7]
```diff
@@ -2720,6 +2735,19 @@ public Iterable<Holiday> getHolidaysForSchedule(Long scheduleId) {
             handleList(Holiday.class, "holidays")));
     }
 
+    public MediaResponse getGuideMedias(Map<String, Object> filterParameters, String pageAfter, int perPage, Sorting sorting) {
+        Map<String, Object> pageParameters = new HashMap<>();
+        pageParameters.put("after", pageAfter);
+        pageParameters.put("size", perPage);
+
+        TemplateUri tmpl = tmpl("/guide/medias{?filter*,page*,sort}")
```
Comment (PavelLoparev) [inline src/main/java/org/zendesk/client/v2/Zendesk.java:2743]: It should be `filter[name]=...`, `page[after]=...` etc. Are you sure this code will produce `filter[name]` query params?

Comment (dimitrystd) [reply src/main/java/org/zendesk/client/v2/Zendesk.java:2743]: Probably this magic will happen in `setGroupParameters`.
But I believe a simple unit test will be the best here. Confirm that the string will be properly formated and also confirm your concern about parameters escaping.

File: src/main/java/org/zendesk/client/v2/Zendesk.java (hunk 6 above)
Comment (sl-pkryshtop) [reply src/main/java/org/zendesk/client/v2/Zendesk.java:2488]: I tested both variants, there are no differences. I remain variant specified in [doc](https://developer.zendesk.com/api-reference/help_center/help-center-api/article_attachments/#create-article-attachment):

File: src/main/java/org/zendesk/client/v2/Zendesk.java (hunk 7 above)
Comment (sl-pkryshtop) [reply src/main/java/org/zendesk/client/v2/Zendesk.java:2743]: there is a test for group parameters in TemplateUriTest, also I tested it with real environment in RealSmokeTest.

File: src/main/java/org/zendesk/client/v2/Zendesk.java (hunk 6 above)
Comment (dimitrystd) [reply src/main/java/org/zendesk/client/v2/Zendesk.java:2488]: I believe it's important and I prefer Pavel's URL. There are 2 reasons:
* Their documentation mentions that an attachment will be created for a specific locale. Maybe there is no difference in API behavior now, but it can be changed in the future. Pavel's URL explicitly includes locale, so it's much easier to argue that we do everything correctly.
* There is a similar problem with Article translations (two API endpoints). But when you have to deal with ZD support, they are picky about every little thing, so it would be good to use the correct URL.

Comment (sl-pkryshtop) [reply src/main/java/org/zendesk/client/v2/Zendesk.java:2488]: If you insist, I will change.
But it isn't a good way to use an undocumented endpoint.
